### PR TITLE
Feat UI metrics search

### LINKED ISSUE #9628: Search field in MetricExplorer
## Proposal
**Use case. Why is this important?**
I went through some early parts of Getting Started guides. When using react UI, navigating to /graph and opening MetricsExplorer modal, there are a lot of metrics to scroll through.

As a user it would be helpful to have a search input to filter the list of metrics with, for ad-hoc analysis.

### PR DESCRIPTION
**Pull request by mtfoley**
This PR resolves #9628 by adding search field for metrics explorer in the React UI.

This PR also adds UI tests for MetricsExplorer component as well.

Screen Shot 2021-10-30 at 11 07 48 PM

I tried to `--signoff` option but I'm not sure I did it correctly.

### CODE REVIEW

**Comment by hanjm** [comment]
The fist commit is not sign, Try [git rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase) and force push this branch 😄 
```
git rebase -i 93d65de1440e3110a6b5bc3199164f02661a9259
```

File: web/ui/react-app/src/pages/graph/MetricsExplorer.tsx [hunk 1]
```diff
@@ -1,33 +1,48 @@
...
-          {this.props.metrics.map((metric) => (
-            <p key={metric} className="metric" onClick={this.handleMetricClick.bind(this, metric)}>
-              {metric}
-            </p>
-          ))}
+          <Input placeholder="Search" value={this.state.searchTerm} type="text" onChange={this.handleSearchTerm} />
+          {this.props.metrics
+            .filter((metric) => {
+              if (this.state.searchTerm === '') return true;
+              return metric.toLowerCase().includes(this.state.searchTerm.toLowerCase());
```
Comment (juliusv) [inline web/ui/react-app/src/pages/graph/MetricsExplorer.tsx:39]: Thanks! We already have search+filtering like this for flags on the /flags page (https://github.com/prometheus/prometheus/blob/143fce73cdc956b577e835a31e785244fc8a3551/web/ui/react-app/src/pages/flags/Flags.tsx), and I think we should keep the various filtering functionalities similar in the way they work. We use `@nexucis/fuzzy` for fuzzy searching + highlighting on the flags page, which I would recommend using here as well to keep things consistent.

Comment (mtfoley) [reply web/ui/react-app/src/pages/graph/MetricsExplorer.tsx]: Took a crack at it, hope that's more like it!

File: web/ui/react-app/src/pages/graph/MetricsExplorer.test.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    expect(metrics).toHaveLength(1);
+  });
+
+  it('handles click on metric', () => {
+    const metric = metricsExplorer.find('.metric').at(0);
+    metric.simulate('click');
+    expect(metricsExplorerProps.insertAtCursor).toHaveBeenCalled();
+  });
+
+  /*  it('executes the query when clicking the execute button', () => {
```
Comment (juliusv) [inline web/ui/react-app/src/pages/graph/MetricsExplorer.test.tsx:47]: This commented-out bit of code should probably be removed?

File: web/ui/react-app/src/pages/graph/MetricsExplorer.tsx [hunk 3]
```diff
@@ -1,33 +1,57 @@
...
 
 interface Props {
   show: boolean;
   updateShow(show: boolean): void;
   metrics: string[];
   insertAtCursor(value: string): void;
 }
-
-class MetricsExplorer extends Component<Props> {
+type MetricsExplorerState = {
```
Comment (juliusv) [inline web/ui/react-app/src/pages/graph/MetricsExplorer.tsx:12]: Style nit: add empty lines before and after such a top-level definition (so one between lines 11+12 and one between lines 14+15).

File: web/ui/react-app/src/pages/graph/MetricsExplorer.tsx [hunk 4]
```diff
@@ -1,33 +1,57 @@
-import React, { Component } from 'react';
-import { Modal, ModalBody, ModalHeader } from 'reactstrap';
+import React, { Component, ChangeEvent } from 'react';
+import { Modal, ModalBody, ModalHeader, Input } from 'reactstrap';
+import { Fuzzy, FuzzyResult } from '@nexucis/fuzzy';
+const fuz = new Fuzzy({ pre: '<strong>', post: '</strong>', shouldSort: true });
```
Comment (juliusv) [inline web/ui/react-app/src/pages/graph/MetricsExplorer.tsx:4]: Style nit: separate this line from the `import`s above with an empty line.

**Review by juliusv — COMMENTED** [review]
Looks great except for a few nits, thanks! :)

File: web/ui/react-app/src/pages/graph/MetricsExplorer.test.tsx (hunk 2 above)
Comment (mtfoley) [reply web/ui/react-app/src/pages/graph/MetricsExplorer.test.tsx]: it had been copied over from an adjacent test file. removed in last commit

File: web/ui/react-app/src/pages/graph/MetricsExplorer.tsx (hunk 3 above)
Comment (mtfoley) [reply web/ui/react-app/src/pages/graph/MetricsExplorer.tsx:14]: should be fixed in last commit

File: web/ui/react-app/src/pages/graph/MetricsExplorer.tsx (hunk 4 above)
Comment (mtfoley) [reply web/ui/react-app/src/pages/graph/MetricsExplorer.tsx:5]: should be fixed in last commit.

File: web/ui/react-app/src/pages/graph/MetricsExplorer.tsx [hunk 5]
```diff
@@ -8,26 +11,50 @@ interface Props {
   insertAtCursor(value: string): void;
 }
 
-class MetricsExplorer extends Component<Props> {
+type MetricsExplorerState = {
+  searchTerm: string;
+};
+
+class MetricsExplorer extends Component<Props, MetricsExplorerState> {
```
[the remark this answers is not in the document]
Comment (mtfoley) [reply web/ui/react-app/src/pages/graph/MetricsExplorer.tsx]: should be fixed on last commit

**Comment by juliusv** [comment]
:+1: Thanks for bearing with me, looks good now :)

**Pull request merged**